kbcstorage/files.py

```diff
@@ -0,0 +1,224 @@
+
+        Returns:
+            file_id (str): Id of the created file
+
+        Raises:
+            requests.HTTPError: If the API request fails.
+        """
+        if not os.path.exists(file_path) or not os.path.isfile(file_path):
+            raise ValueError("File " + file_path + " does not exist")
+        file_name = os.path.basename(file_path)
+        size = os.path.getsize(file_path)
+        file_resource = self.prepare_upload(file_name, size, tags, is_public,
+                                            is_permanent, is_encrypted,
+                                            is_sliced, do_notify, True)
+        upload_params = file_resource['uploadParams']
+        key_id = upload_params['credentials']['AccessKeyId']
+        key = upload_params['credentials']['SecretAccessKey']
+        token = upload_params['credentials']['SessionToken']
+        s3 = boto3.resource('s3', aws_access_key_id=key_id,
+                            aws_secret_access_key=key,
+                            aws_session_token=token,
+                            region_name=file_resource['region'])
+
+        s3_object = s3.Object(bucket_name=upload_params['bucket'],
+                              key=upload_params['key'])
+        disposition = 'attachment; filename={};'.format(file_resource['name'])
+        with open(file_path, mode='rb') as file:
+            if is_encrypted:
+                encryption = upload_params['x-amz-server-side-encryption']
+                s3_object.put(ACL=upload_params['acl'], Body=file,
+                              ContentDisposition=disposition,
+                              ServerSideEncryption=encryption)
+            else:
+                s3_object.put(ACL=upload_params['acl'], Body=file,
+                              ContentDisposition=disposition)
+        return file_resource['id']
+
+    def prepare_upload(self, name, size_bytes=None, tags=None, is_public=False,
+                       is_permanent=False, is_encrypted=True,
+                       is_sliced=False, do_notify=False,
+                       federation_token=True):
+        """
+        Prepare a file resource for a new file
+
+        Args:
+            name (str): The new file name (only alphanumeric and underscores)
+            size_bytes (int): Size of the file
+            tags (list): Array of tags
+            is_public (bool): File is public
+            is_permanent (bool): File is permanent
+            is_encrypted (bool): File is encrypted
+            is_sliced (bool): File is sliced
+            do_notify (bool): Notify members of project that file was uploaded
+            federation_token (bool): Obtain AWS federation token
+
+        Returns:
+            response_body: The parsed json from the HTTP response.
+
+        Raises:
+            requests.HTTPError: If the API request fails.
+        """
+        url = '{}/{}'.format(self.base_url, 'prepare')
+        headers = {

```

Comment:
`requests` handles the headers automatically

```
>>>import requests
>>>resp = requests.post('https://www.httpbin.org/post', data={'foo': 'bar'})
>>>print(resp.request.headers)
{'User-Agent': 'python-requests/2.12.4', 
 'Connection': 'keep-alive', 
 'Accept-Encoding': 'gzip, deflate',
 'Accept': '*/*', 'Content-Length': '7', 
 'Content-Type': 'application/x-www-form-urlencoded'}
```

In the PR mentioned in the previous comment (to be published soon; unfortunately I almost finished it few minutes before this massive PR was made) I refactor this. Also I refactored the authentication headers handling into the base endpoint http methods `Endpoint._get()`. The refactor will enable passing of custom headers aswell, but avoids this tedious rewriting.

Reply:
i made a separate issue for this https://github.com/keboola/sapi-python-client/issues/30 as there is a slight catch (like everywhere), definitely a job for next PR